lib/proxy.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+const _ = require('lodash');
+const defaultOptions = {
+  models: new Map()
+};
+
+class ModelProxy {
+
+  /**
+   * @param {object} opts
+   * @param {Map} opts.models - Map of the model objects
+   * @param {string} opts.primary - name of the primary model
+   * @param {string|boolean} opts.secondary - name of the secondary model or false if no secondary
+   * @param {object} opts.logger - log object (log, info, error)
+   */
+  constructor(opts) {
+    const options = _.merge({}, defaultOptions, opts || {});
+
+    // Set primary model on options
+    options.primaryModel = setupModel(options.primary, options.models);
+
+    // Conditionally set secondary model on options
+    options.secondaryModel = setupModel(options.secondary, options.models);
+
+    // Fail if primary and secondary are accidentally the same
+    if (isEqual(options.primary, options.secondary)) {
+      throw new Error('Primary and secondary models cannot be the same');
+    }
+
+    // Get methods from own and inherited enumerable properties on primaryModel
+    const methodsToProxy = _.functionsIn(options.primaryModel);
+
+    // Loop over primary methods and attach proxied version to this
+    _.forEach(methodsToProxy, (methodName) => {
+      this[methodName] = proxify(methodName, options);
+    });
+  }
+}
+
+module.exports = ModelProxy;
+
+/**
+ * Test if 2 items are equal, but only if they are NOT both undefined
+ *
+ * @param {*} a
+ * @param {*} b
+ * @return {boolean}
+ */
+function isEqual(a, b) {
+  if (typeof a === 'undefined' && typeof b === 'undefined') {
+    return false;
+  }
+  return a === b;
+}
+
+/**
+ * Get a model out of the model map, and throw an error if it doesn't exist
+ *
+ * @param {string} name - name of the model to get
+ * @param {Map} models - map to get model from
+ *
+ * @return {object} - the extracted model
+ */
+function setupModel(name, models) {
+  const model = models.get(name);
+
+  if (name && !model) {
+    throw new Error(`Must include model named "${name}" in models passed to proxy`);
+  }
+
+  return model;
+}
+
+/**
+ * Create a new function that calls primary model method, and conditionally, the secondary
+ *
+ * @param {string} method - name of the method to proxy
+ * @param {object} options
+ * @param {string} options.primary - name of the primary model
+ * @param {object} options.primaryModel - primary model object
+ * @param {string|boolean} options.secondary - name of the secondary model or false if no secondary
+ * @param {object|undefined} options.secondaryModel - secondary model object or undefined if no secondary
+ * @param {object} options.logger - object with log, info, error log functions
+ *
+ * @return {function} new proxied function
+ */
+function proxify(method, options) {
+  return function() {
+    const args = (arguments.length === 1) ? [arguments[0]] : Array.apply(null, arguments);
+
+    return callMethod(options.primaryModel, method, args).then((result) => {
```

Comment:
Any reason to not call these in parallel?  Just wondering if we desired some of the side-effects here (i.e., if primary fails then secondary won't be called)